udp/client/conn.go

```diff
@@ -609,34 +688,14 @@
 	return cc.sequence.Add(1)
 }
 
-func (cc *Conn) responseMsgCacheID(msgID int32) string {
-	return fmt.Sprintf("resp-%v-%d", cc.RemoteAddr(), msgID)
+// getResponseFromCache gets a message from the response message cache.
+func (cc *Conn) getResponseFromCache(mid int32, resp *pool.Message) (bool, error) {
+	return cc.responseMsgCache.Load(strconv.Itoa(int(mid)), resp)
 }
 
+// addResponseToCache adds a message to the response message cache.
 func (cc *Conn) addResponseToCache(resp *pool.Message) error {
-	marshaledResp, err := resp.MarshalWithEncoder(coder.DefaultCoder)
-	if err != nil {
-		return err
-	}
-	cacheMsg := make([]byte, len(marshaledResp))
-	copy(cacheMsg, marshaledResp)
-	cc.responseMsgCache.LoadOrStore(cc.responseMsgCacheID(resp.MessageID()), cache.NewElement(cacheMsg, time.Now().Add(ExchangeLifetime), nil))
-	return nil
-}
-
-func (cc *Conn) getResponseFromCache(mid int32, resp *pool.Message) (bool, error) {
-	cachedResp := cc.responseMsgCache.Load(cc.responseMsgCacheID(mid))
-	if cachedResp == nil {
-		return false, nil
-	}
-	if rawMsg := cachedResp.Data(); len(rawMsg) > 0 {
-		_, err := resp.UnmarshalWithDecoder(coder.DefaultCoder, rawMsg)
-		if err != nil {
-			return false, err
-		}
-		return true, nil
-	}
-	return false, nil
+	return cc.responseMsgCache.Store(strconv.Itoa(int(resp.MessageID())), resp)
```

Comment:
I agree. Collisions could occur – could you please use responseMsgCacheID(resp.MessageID()) as the key?

Reply:
@jkralik I intentionally made this update because, if using DTLS 1.2 Connection IDs, the underlying remote IP address / port can change and the connection can continue. Items in the cache would then no longer be accessible because we would attempt to access with a different key. In the event that we are not using DTLS, or using DTLS without Connection IDs, I don't believe that `Conn`s should be used across multiple endpoints, and I believe there are issues other than just the cache keys here if they are. Is there a scenario in which a `Conn` is being used across endpoints, thus meaning that a `responseMsgCache` would not be local to a single endpoint?

Reply:
Ok. I don’t remember an issue, but I ran tests on all our main products, and it seems the issue isn’t there. So tomorrow, when all tests pass, I’ll merge it.